### PR TITLE
Admin can send rich text emails to receivers in a list generated by script #2948

### PR DESCRIPTION
**Pull request by ymoctavia**
for #2948

### CODE REVIEW

**Comment by ymoctavia** [comment]
the task queue deleting is done in #3076

**Comment by damithc** [comment]
Rename the issue/PR to a better name.

File: src/main/java/teammates/client/scripts/AdminEmailListGenerator.java [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+import teammates.common.util.Const;
+import teammates.common.util.TimeHelper;
+import teammates.storage.entity.Account;
+import teammates.storage.entity.Course;
+import teammates.storage.entity.Instructor;
+import teammates.storage.entity.Student;
+
+import teammates.common.exception.InvalidParametersException;
+
+public class AdminEmailListGenerator extends RemoteApiClient {
```
Comment (damithc) [inline src/main/java/teammates/client/scripts/AdminEmailListGenerator.java]: Add a header comment for the class

File: src/main/java/teammates/client/scripts/AdminEmailListGenerator.java [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+           System.out.print("Date is not valid. Please Re-enter date.\n\n");
+       } else {
+           System.out.print("Date Entered is valid.\n\n");
+       }
+       
+       return isDateValid;
+        
+    }
+    
+    private int getMaxNumOfDayForMonth(int month, int year){
```
Comment (damithc) [inline src/main/java/teammates/client/scripts/AdminEmailListGenerator.java]: If Java doesn't have built-in libraries for this, may be use an external library such as jodatime or natty?

File: src/main/java/teammates/client/scripts/AdminEmailListGenerator.java [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+            } else {
+                throw new InvalidParametersException("Date format error");
+            }
+        } catch(Exception e) {
+            throw new InvalidParametersException("Date format error");
+        }
+        
+    }
+    
+    private void printToFile() {
```
Comment (damithc) [inline src/main/java/teammates/client/scripts/AdminEmailListGenerator.java]: Apply SLAP more here. This method is long and nesting level is deep.

File: src/main/java/teammates/client/scripts/AdminEmailListGenerator.java [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+        Course course = getCourseEntity(instructor.getCourseId());
+        
+        if(course != null){
+            CourseIdToCreatedDateMap.put(instructor.getCourseId(), course.getCreatedAt());
+            return course.getCreatedAt();
+        }
+        
+        return null;
+        
+}
```
Comment (damithc) [inline src/main/java/teammates/client/scripts/AdminEmailListGenerator.java]: This `}` doesn't seem to be indented correctly.

File: src/main/java/teammates/common/util/Const.java [hunk 5]
```diff
@@ -24,6 +24,9 @@
         /** This is the limit given to Blobstore API, beyond which an ugly error page is shown */
         public static final long MAX_PROFILE_PIC_LIMIT_FOR_BLOBSTOREAPI = 11000000;
         
+        /** This is the limit given to Blobstore API, beyond which an ugly error page is shown */
+        public static final long MAX_ADMIN_EMAIL_FILE_LIMIT_FOR_BLOBSTOREAPI = 11000000;
```
Comment (damithc) [inline src/main/java/teammates/common/util/Const.java]: `BLOBSTORE_API`?

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 6]
```diff
@@ -87,7 +93,29 @@
...
     // ////////////////////////////////////////////////////////////////////////
-
+    
+    
+    /*
+     * ======================================================================= 
+     * Field: Email Subject
+     */
+    private static final String EMAIL_SUBJECT_FIELD_NAME = "email subject";
+    // one more than longest official nationality name
```
Comment (damithc) [inline src/main/java/teammates/common/util/FieldValidator.java]: What's the connection to the 'official nationality name'?

File: src/main/java/teammates/logic/automated/AdminEmailPrepareTaskQueueWorkerServlet.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+            
+            String indexOfEmailListToResumeAsString = HttpRequestHelper.getValueFromRequestParameterMap(req, ParamsNames.ADMIN_GROUP_RECEIVER_EMAIL_LIST_INDEX);
+            String indexOfEmailToResumeAsString = HttpRequestHelper.getValueFromRequestParameterMap(req, ParamsNames.ADMIN_GROUP_RECEIVER_EMAIL_INDEX);
+            
+            int indexOfEmailListToResume = indexOfEmailListToResumeAsString == null ? 0 : Integer.parseInt(indexOfEmailListToResumeAsString);
+            int indexOfEmailToResume = indexOfEmailToResumeAsString == null ? 0 : Integer.parseInt(indexOfEmailToResumeAsString);
+            
+            
+            
+
```
Comment (damithc) [inline src/main/java/teammates/logic/automated/AdminEmailPrepareTaskQueueWorkerServlet.java]: Too many blank lines?

File: src/main/java/teammates/logic/automated/AdminEmailPrepareTaskQueueWorkerServlet.java [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    
+    private List<List<String>> getReceiverList(String listFileKey, int size) 
+            throws IOException {
+        
+        Assumption.assertNotNull(listFileKey);
+        
+       
+        BlobKey blobKey = new BlobKey(listFileKey);
+        
+
```
Comment (damithc) [inline src/main/java/teammates/logic/automated/AdminEmailPrepareTaskQueueWorkerServlet.java]: too many blank lines here too

File: src/main/java/teammates/logic/publicresource/PublicImageServlet.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
+package teammates.logic.publicresource;
+
```
Comment (damithc) [inline src/main/java/teammates/logic/publicresource/PublicImageServlet.java]: Add header comments to all non-trivial classes

File: src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+        String emailId = getRequestParamValue(Const.ParamsNames.ADMIN_EMAIL_ID);
+        
+        addressReceiver.add(addressReceiverListString);
+        
+        if(groupReceiverListFileKey != null && !groupReceiverListFileKey.isEmpty()){
+            groupReceiver.add(groupReceiverListFileKey);
+        }
+        
+        boolean isNewDraft = emailId == null;
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java]: `isNewDraft = (emailId == null)`

File: src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+           
+            createAndSaveNewDraft(subject, addressReceiver, groupReceiver, emailContent);
+        } else {                     
+            AdminEmailAttributes previousDraft = logic.getAdminEmailById(emailId);
+            
+            if(previousDraft == null){
+                createAndSaveNewDraft(subject, addressReceiver, groupReceiver, emailContent);
+            } else {
+                updatePreviousEmailDraft(previousDraft.getEmailId(), subject, addressReceiver, groupReceiver, emailContent);
+            }
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java]: The logic here is not clear immediately

File: src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+           if(error != null && !error.isEmpty()){
+               isError = true;
+               statusToUser.add(error);
+               throw new InvalidParametersException("<strong>Email Format Error</strong>");
+           }
+       }
+       
+    }
+    
+    private void checkGroupReceiverListFile(String listFileKey) throws IOException {
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java]: Not immediately clear what this method does. To this method and other methods that are long, 1. try to simplify using SLAP. 2. add more comments.

File: src/main/webapp/WEB-INF/web.xml [hunk 13]
```diff
@@ -274,6 +294,17 @@ http://java.sun.com/xml/ns/javaee/web-app_2_5.xsd"
...
+    
+    <servlet>
+        <servlet-name>PublicImageServlet</servlet-name>
+        <servlet-class>teammates.logic.publicresource.PublicImageServlet</servlet-class>
+    </servlet>
+    
+     <servlet-mapping>
+        <servlet-name>PublicImageServlet</servlet-name>
+        <url-pattern>/public/publicEmailImageServe</url-pattern>
+    </servlet-mapping>
```
Comment (damithc) [inline src/main/webapp/WEB-INF/web.xml]: Make sure new actions have a correct security setting.

**Comment by damithc** [comment]
Some tweaks suggested.

**Comment by ymoctavia** [comment]
Professor @damithc , this is updated ,too.

File: .settings/com.google.appengine.eclipse.core.prefs [hunk 14]
```diff
@@ -3,6 +3,7 @@ filesCopiedToWebInfLib=appengine-api-1.0-sdk-1.9.4.jar|appengine-api-labs.jar|ap
 gaeDatanucleusVersion=v1
 gaeDeployDialogSettings=
 gaeIsEclipseDefaultInstPath=true
+gaeLaunchAppInBrowser=true
```
Comment (damithc) [inline .settings/com.google.appengine.eclipse.core.prefs]: not relevant

**Comment by damithc** [comment]
Add header comments to classes.
Increase SLAP where possible.
Make sure this doesn't break other admin tests.
Can merge after that.

**Pull request merged**